### PR TITLE
Pay plugin: Fix the traversal order in `find_worst_channel`

### PR DESCRIPTION
**Pull request by trueptolemy**
Related with my ['question' comment](https://github.com/ElementsProject/lightning/issues/2665#issuecomment-531479284)

I guess the original idea is to use the reverse order to traverse the entire route collection to find the worst channel/hop. I think it's the reason of the 'abnormal performance' in `test_pay_limits` if I'm right...

But the reverse order is hard to implement for our json interface, so I fix the logic with `json_for_each_arr`.

### CODE REVIEW

File: plugins/pay.c [hunk 1]
```diff
@@ -454,19 +454,38 @@ static const jsmntok_t *find_worst_channel(const char *buf,
...
+	const jsmntok_t *worst = NULL, *t, *t_prev = NULL;
 	size_t i;
 
 	json_for_each_arr(i, t, route) {
 		u64 val;
+		/* If there's only 2 hops, the worst channel must be the first one. */
+		if (route->size <= 2) {
+			worst = t;
+			break;
+		}
```
Comment (rustyrussell) [inline plugins/pay.c]: Don't special-case this, I don't think it needs it?

File: plugins/pay.c [hunk 2]
```diff
@@ -454,19 +454,38 @@ static const jsmntok_t *find_worst_channel(const char *buf,
...
+		 * Just store the info and continue. */
+		if (!i) {
+			prev = val;
+			t_prev = t;
+			continue;
+		}
+
+		/* the final hop */
+		if (i + 1 == route->size)
+			assert(val == final);
```
Comment (rustyrussell) [inline plugins/pay.c]: We should just drop the final arg, it's now redundant.

File: plugins/pay.c (hunk 1 above)
Comment (trueptolemy) [reply plugins/pay.c]: Yes...there's no need.

File: plugins/pay.c (hunk 2 above)
Comment (trueptolemy) [reply plugins/pay.c]: Yes, I remove this, and also remove the `final` parameter in this interface.

**Comment by trueptolemy** [comment]
@rustyrussell I'm also not sure about the sencond commit https://github.com/ElementsProject/lightning/pull/3061/commits/07f506f17dea7a987c696c45ae77772eacd6ecc3 :
Is it ok if I add more than one `destructor` for a `tal` point?
Here `master` as the `struct daemon_conn` already has a `destructor`, and I add another special `destructor` on it.

**Comment by trueptolemy** [comment]
Thank you! Very soon! :-)

**Comment by rustyrussell** [comment]
> @rustyrussell I'm also not sure about the sencond commit [07f506f](https://github.com/ElementsProject/lightning/commit/07f506f17dea7a987c696c45ae77772eacd6ecc3) :
> Is it ok if I add more than one `destructor` for a `tal` pointer?
> Here `master` as the `struct daemon_conn` already has a `destructor`, and I add another special `destructor` on it.

Yep, completely legal.

**Pull request merged**